ktor-server/ktor-server-core/jvm/src/io/ktor/server/application/plugins/api/ServerPlugin.kt

```diff
@@ -206,58 +236,87 @@ public abstract class ServerPlugin<Configuration : Any> private constructor(
+            createConfiguration: () -> Configuration,
+            body: ApplicationPlugin<Configuration>.() -> Unit
+        ): ServerApplicationPluginFactory<Configuration> = object : ServerApplicationPluginFactory<Configuration>() {
             override val key: AttributeKey<ServerPlugin<Configuration>> = AttributeKey(name)
 
             override fun install(
-                pipeline: ApplicationCallPipeline,
+                pipeline: Application,
                 configure: Configuration.() -> Unit
-            ): ServerPlugin<Configuration> {
-                val config = createConfiguration(pipeline)
+            ): ApplicationPlugin<Configuration> {
+                val config = createConfiguration()
                 config.configure()
 
                 val self = this
-                val pluginInstance = object : ServerPlugin<Configuration>(self) {
-                    override val pipeline: ApplicationCallPipeline
-                        get() = pipeline
-                    override val pluginConfig: Configuration
-                        get() = config
-                }
-
-                pluginInstance.apply(body)
-
-                pluginInstance.pipelineHandlers.forEach { handle ->
-                    handle(pipeline)
+                val pluginInstance = object : ApplicationPlugin<Configuration>(self) {
+                    override val pipeline: ApplicationCallPipeline = pipeline
+                    override val pluginConfig: Configuration = config
                 }
+                pluginInstance.setupPlugin(body)
+                return pluginInstance
+            }
+        }
 
-                pluginInstance.callInterceptions.forEach {
-                    it.action(pipeline)
-                }
+        /**
+         * Creates a [ServerRoutingScopedPluginFactory] that can be installed into [io.ktor.server.routing.Routing].
+         *
+         * @param name A name of your new plugin that will be used if you need find an instance of
+         * your plugin when it is installed to an [io.ktor.server.routing.Routing].
+         * @param createConfiguration Defines how the initial [Configuration] of your new plugin can be created. Please
+         * note that it may be modified later when a user of your plugin calls [install].
+         * @param body Allows you to define handlers ([onCall], [onCallReceive], [onCallRespond] and so on) that
+         * can modify the behaviour of an [Application] where your plugin is installed.
+         *
+         * Usage example:
+         * ```
+         * val MyPlugin = createPlugin("MyPlugin") {
+         *      // This block will be executed when you call install(MyPlugin)
+         *
+         *      onCall { call ->
+         *          // Prints requested URL each time your application receives a call:
+         *          println(call.request.uri)
+         *      }
+         * }
+         *
+         * route("a") {
+         *   install(MyPlugin)
```

Comment:
I would add some example describing the semantics of the re-installation into a different subroute

Reply:
cc: @andreyaksenov